### ISSUE TITLE
0.100.7 SwitchSequential bug?

### ISSUE THREAD

**Issue by williamlzw** [dotnet/TorchSharp#1132]
public class ResidualBlock : Module<Tensor, Tensor, Tensor>
      {
          private readonly GroupNorm groupnorm_feature;
          private readonly Conv2d conv_feature;
          private readonly Linear linear_time;
          private readonly GroupNorm groupnorm_merged;
          private readonly Conv2d conv_merged;
          private readonly Module<Tensor, Tensor> residual_layer;
          private readonly SiLU silu;
      
          public ResidualBlock(long in_channels, long out_channels, long n_time = 1280) : base("ResidualBlock")
          {
              groupnorm_feature = GroupNorm(32, in_channels);
              conv_feature = Conv2d(in_channels, out_channels, kernelSize: 3, padding: 1);
              linear_time = Linear(n_time, out_channels);
              groupnorm_merged = GroupNorm(32, out_channels);
              conv_merged = Conv2d(out_channels, out_channels, kernelSize: 3, padding: 1);
              silu = SiLU();
              if (in_channels == out_channels)
              {
                  residual_layer = nn.Identity();
              }
              else
              {
                  residual_layer = Conv2d(in_channels, out_channels, kernelSize: 1);
              }
              RegisterComponents();
          }
      
          public override Tensor forward(Tensor feature, Tensor time)
          {
              using var _ = NewDisposeScope();
              //Console.WriteLine($"ResidualBlock.input.feature {string.Join(",", feature.shape)}");
              var residue = feature;
              feature = groupnorm_feature.forward(feature);
              feature = silu.forward(feature);
              feature = conv_feature.forward(feature);
      
              time = silu.forward(time);
              time = linear_time.forward(time);
      
              var merged = feature + time.unsqueeze(-1).unsqueeze(-1);
             
              merged = groupnorm_merged.forward(merged);
              merged = silu.forward(merged);
              merged = conv_merged.forward(merged);
              var output = merged + residual_layer.forward(residue);
              //Console.WriteLine($"ResidualBlock.output {string.Join(",", output.shape)}");
              return output.MoveToOuterDisposeScope();
          }
      }
      
      public class AttentionBlock : Module<Tensor, Tensor, Tensor>
      {
          private readonly GroupNorm groupnorm;
          private readonly Conv2d conv_input;
          private readonly LayerNorm layernorm_1;
          private readonly SelfAttention attention_1;
          private readonly LayerNorm layernorm_2;
          private readonly CrossAttention attention_2;
          private readonly LayerNorm layernorm_3;
          private readonly Linear linear_geglu_1;
          private readonly Linear linear_geglu_2;
          private readonly Conv2d conv_output;
      
          public AttentionBlock(long n_head, long n_embd, long d_context = 768) : base("AttentionBlock")
          {
              var channels = n_head * n_embd;
              groupnorm = GroupNorm(32, channels);
              conv_input = Conv2d(channels, channels, kernelSize: 1);
              layernorm_1 = LayerNorm(channels);
              attention_1 = new SelfAttention(n_head, channels, in_proj_bias: false);
              layernorm_2 = LayerNorm(channels);
              attention_2 = new CrossAttention(n_head, channels, d_context, in_proj_bias: false);
              layernorm_3 = LayerNorm(channels);
              linear_geglu_1 = Linear(channels, 4 * channels * 2);
              linear_geglu_2 = Linear(4 * channels, channels);
              conv_output = Conv2d(channels, channels, kernelSize: 1);
              RegisterComponents();
          }
      
          public override Tensor forward(Tensor x, Tensor context)
          {
              using var _ = NewDisposeScope();
             // Console.WriteLine($"AttentionBlock.input.x {string.Join(",", x.shape)}");
              var residue_long = x;
              x = groupnorm.forward(x);
              x = conv_input.forward(x);
              var n = x.shape[0];
              var c = x.shape[1];
              var h = x.shape[2];
              var w = x.shape[3];
              x = x.view(new long[] { n, c, h * w });
              x = x.transpose(-1, -2);
              var residue_short = x;
              x = layernorm_1.forward(x);
             // Console.WriteLine($"AttentionBlock.layernorm_1.output: {string.Join(",", x.shape)}");
              x = attention_1.forward(x);
              x += residue_short;
              residue_short = x;
              x = layernorm_2.forward(x);
              x = attention_2.forward(x, context);
              x+= residue_short;
              residue_short = x;
              x = layernorm_3.forward(x);
              var ret = linear_geglu_1.forward(x).chunk(2, -1);
              x = ret[0];
              var gate = ret[1];
              x = x * torch.nn.functional.gelu(gate);
              x = linear_geglu_2.forward(x);
              x += residue_short;
              x = x.transpose(-1, -2);
              x = x.view(new long[] { n, c, h, w });
              var output = conv_output.forward(x) + residue_long;
             // Console.WriteLine($"AttentionBlock.last.output: {string.Join(",", output.shape)}");
              return output.MoveToOuterDisposeScope();
          }
      }
      
      public class Upsample : Module<Tensor, Tensor>
      {
          private readonly Conv2d conv;
          public Upsample(long channels) : base("Upsample")
          {
              conv = Conv2d(channels, channels, kernelSize: 3, padding: 1);
              RegisterComponents();
          }
      
          public override Tensor forward(Tensor x)
          {
              using var _ = NewDisposeScope();
              x = torch.nn.functional.interpolate(x, scale_factor: new double[] { 2, 2 });
              var output = conv.forward(x);
              return output.MoveToOuterDisposeScope();
          }
      }
      
      class SwitchSequential : Sequential<Tensor, Tensor, Tensor, Tensor>
      {
          internal SwitchSequential(params (string name, torch.nn.Module)[] modules) : base(modules)
          {
          }
      
          internal SwitchSequential(params torch.nn.Module[] modules) : base(modules)
          {
          }
      
          public override torch.Tensor forward(torch.Tensor x, torch.Tensor context, torch.Tensor time)
          {
              using var _ = torch.NewDisposeScope();
              foreach (var layer in modules())
              {
                  Console.WriteLine($"SwitchSequential.layer.type :{layer.GetType()} ,{string.Join(",", x.shape)}");
                  switch (layer)
                  {
                      case AttentionBlock abl:
                          x = abl.call(x, context);
                          break;
                      case ResidualBlock rbl:
                          x = rbl.call(x, time);
                          break;
                      case torch.nn.Module<torch.Tensor, torch.Tensor> m:
                          x = m.call(x);
                          break;
                  }
              }
              return x.MoveToOuterDisposeScope();
          }
      }

       public static void test()
       {
           var encoders = new ModuleList<SwitchSequential>(
                   new SwitchSequential(nn.Conv2d(4, 320, 3, padding: 1)),
                   new SwitchSequential(new ResidualBlock(320, 320), new AttentionBlock(8, 40)),
                   new SwitchSequential(new ResidualBlock(320, 320), new AttentionBlock(8, 40)));
           encoders.eval();
           var x = torch.randn(2, 4, 64, 64);
           var context = torch.randn(2, 77, 768);
           var time = torch.randn(1, 1280);
           var input = x.alias();
           foreach (var layers in encoders)
           {
               Console.WriteLine($"encoders.layers.type {layers.GetType()}");
               input = layers.forward(input, context, time);
              // Console.WriteLine($"UNet1.encoders  {string.Join(",", x.shape)}");
           }
       }

  ================================
console print
encoders.layers.type StableDiffusionTorchSharp.SwitchSequential
SwitchSequential.layer.type :TorchSharp.Modules.Conv2d ,2,4,64,64
encoders.layers.type StableDiffusionTorchSharp.SwitchSequential
SwitchSequential.layer.type :StableDiffusionTorchSharp.ResidualBlock ,2,320,64,64
SwitchSequential.layer.type :StableDiffusionTorchSharp.AttentionBlock ,2,320,64,64
SwitchSequential.layer.type :TorchSharp.Modules.GroupNorm ,2,320,64,64//????????????? Where's the extra layer coming from
SwitchSequential.layer.type :TorchSharp.Modules.Conv2d ,2,320,64,64//?????????????Where's the extra layer coming from
SwitchSequential.layer.type :TorchSharp.Modules.Linear ,2,320,64,64//?????????????Where's the extra layer coming from

**Comment by NiklasGustafsson** [comment]
Can you try `children()` instead of `modules()` in the `forward` of `SwitchSequential`?

**Comment by williamlzw** [comment]
The code passed.  
   class SwitchSequential : Sequential<Tensor, Tensor, Tensor, Tensor>
    {
        internal SwitchSequential(params (string name, torch.nn.Module)[] modules) : base(modules)
        {
        }
    
        internal SwitchSequential(params torch.nn.Module[] modules) : base(modules)
        {
        }
    
        public override torch.Tensor forward(torch.Tensor x, torch.Tensor context, torch.Tensor time)
        {
            using var _ = torch.NewDisposeScope();
            x = x.alias();
            foreach (var layer in children())
            {
                switch (layer)
                {
                    case AttentionBlock abl:
                        x = abl.call(x, context);
                        break;
                    case ResidualBlock rbl:
                        x = rbl.call(x, time);
                        break;
                    case torch.nn.Module<torch.Tensor, torch.Tensor> m:
                        x = m.call(x);
                        break;
                }
            }
            return x.MoveToOuterDisposeScope();
        }
    }
